### PR TITLE
Migration from ethers to alloy types

### LINKED ISSUE #1681: Migration from `ethers` to `alloy`
**Is your feature request related to a problem? Please describe.**

In the codebase, there are several instances where `ethers` types are still being used, such as:

https://github.com/dojoengine/dojo/blob/91d8dac307e5f09b06cc47e2e5427d930b593a39/crates/katana/rpc/rpc-types/src/account.rs#L1
https://github.com/dojoengine/dojo/blob/91d8dac307e5f09b06cc47e2e5427d930b593a39/crates/katana/primitives/src/transaction.rs#L2
https://github.com/dojoengine/dojo/blob/91d8dac307e5f09b06cc47e2e5427d930b593a39/crates/katana/core/src/service/messaging/ethereum.rs#L9

**Describe the solution you'd like**

Following discussions like [this one](https://github.com/paradigmxyz/reth/issues/4715), among others, numerous repositories are transitioning from `ethers` to `alloy` as a more comprehensive, precise, and modern Rust Ethereum library. Consequently, it appears appropriate to migrate entirely to the equivalent `alloy` library here, eliminating all references to `ethers`.

### PR DESCRIPTION
**Pull request by tcoratger**
This PR handles the migration from ethers to alloy types.

Should close #1681.
Related to DOJ-277.

### CODE REVIEW

File: crates/katana/core/src/service/messaging/ethereum.rs
Comment (glihm) [inline crates/katana/core/src/service/messaging/ethereum.rs:184]: Is there a way to have this method in `snake_case` as mentioned in the other comment?

File: crates/katana/core/src/service/messaging/ethereum.rs
Comment (glihm) [inline crates/katana/core/src/service/messaging/ethereum.rs:23]: Is `snake_case` supported here without any side effect? If yes, we would prefer `snake_case`. :+1:

File: Cargo.toml
Comment (glihm) [inline Cargo.toml:206]: Were those required here? Only Katana and Saya should rely `alloy`.

**Review by glihm — COMMENTED** [review]
Awesome, thank you @tcoratger for this update!
Some minor comments and one question:

Why a specific rev is used? At the moment it still recommended to use git rev instead of `crates.io`?

File: crates/katana/core/src/service/messaging/ethereum.rs
Comment (tcoratger) [reply crates/katana/core/src/service/messaging/ethereum.rs:184]: I've tried to put a snake case renaming constraint in the `sol!` macro but it doesn't seem to work. The functions around this macro are not yet all stable as specified in their repo and so I tell myself that it is surely related.

File: crates/katana/core/src/service/messaging/ethereum.rs
Comment (tcoratger) [reply crates/katana/core/src/service/messaging/ethereum.rs]: Doesn't seem to work as specified here https://github.com/dojoengine/dojo/pull/1693#discussion_r1536872787

File: Cargo.toml
Comment (tcoratger) [reply Cargo.toml:206]: There are not dependencies here but just patch for crate io in general (I believe the patch is only supported at root level) because the actual version provided bu crate.io is not valid as confirmed here: https://github.com/alloy-rs/alloy/issues/350#issuecomment-2010960282

**Comment by glihm** [comment]
The saya test was not always passing, it's fixed on main once #1710 is merged.

**Comment by tcoratger** [comment]
> The saya test was a bit flaky, it's fixed on main once #1710 is merged.

@glihm Let me know when we can merge this one and if we need to do something else here.

**Comment by glihm** [comment]
> > The saya test was a bit flaky, it's fixed on main once #1710 is merged.
> 
> @glihm Let me know when we can merge this one and if we need to do something else here.

Amazing, thank you for all the work done here, will review shortly. :+1:

File: crates/katana/core/Cargo.toml
Comment (kariy) [inline crates/katana/core/Cargo.toml:50]: except for the primitives, can we make the others optional ? considering its only used in `messaging` feature

File: crates/katana/core/src/backend/config.rs
Comment (kariy) [inline crates/katana/core/src/backend/config.rs:3]: we can remove `primitive_types` from `Cargo.toml` now

File: scripts/clippy.sh
Comment (kariy) [inline scripts/clippy.sh:1]: is this included accidentally?

**Review by kariy — CHANGES_REQUESTED** [review]
looks good overall. Seems to me `alloy-primitives` is used in many crates, maybe it's better to have it on workspace level.

File: crates/saya/core/Cargo.toml
Comment (kariy) [inline crates/saya/core/Cargo.toml:41]: i believe this crate doesn't need it
```suggestion
```

**Review by kariy — CHANGES_REQUESTED** [review]

File: crates/saya/core/Cargo.toml
Comment (tcoratger) [reply crates/saya/core/Cargo.toml]: Done

File: crates/katana/core/src/backend/config.rs
Comment (tcoratger) [reply crates/katana/core/src/backend/config.rs:3]: Done

File: crates/katana/core/Cargo.toml
Comment (tcoratger) [reply crates/katana/core/Cargo.toml]: Done

File: scripts/clippy.sh
Comment (tarrencev) [inline scripts/clippy.sh:5]:
```suggestion
```

We shouldn't include this here since it will go out of sync with the rust toolchain specified elsewhere

**Review by kariy — APPROVED** [review]

**Review by glihm — APPROVED** [review]

**Pull request merged**
